utils/liquidity.ts

```diff
@@ -0,0 +1,199 @@
+};
+
+/**
+ * Compute the numbers of outcome tokens that will be sent to the user by the market maker after adding `addedFunds` of collateral.
+ * @param addedFunds - the amount of collateral being added to the market maker as liquidity
+ * @param poolBalances - the market maker's balances of outcome tokens
+ */
+export const calcAddFundingOutcomeTokensReturned = (
+  addedFunds: bigint,
+  poolBalances: bigint[],
+  poolShareSupply?: bigint
+): bigint[] | null => {
+  if (poolShareSupply && poolShareSupply === BigInt(0)) {
+    return null;
+  }
+  const depositAmounts = calcAddFundingDepositedAmounts(addedFunds, poolBalances);
+
+  const sendAmounts = depositAmounts.map(depositAmount => addedFunds - depositAmount);
+  return sendAmounts;
+};
+
+/**
+ * Compute the amount of collateral that can be obtained via merging after the user
+ * removed `removedFunds` of pool shares.
+ * @param removedFunds - the amount of liquidity pool tokens being sent to the market maker in return for underlying outcome tokens
+ * @param poolBalances - the market maker's balances of outcome tokens
+ * @param poolShareSupply - the total supply of liquidity pool tokens
+ */
+export const calcDepositedTokens = (
+  removedFunds: bigint,
+  poolBalances: bigint[],
+  poolShareSupply: bigint
+): bigint => {
+  const sendAmounts = calcRemoveFundingSendAmounts(
+    removedFunds,
+    poolBalances,
+    poolShareSupply
+  );
+  return sendAmounts.reduce((min, amount) => (amount < min ? amount : min));
+};
+
+/**
+ * Compute the numbers of outcome tokens that will be added to the market maker after adding `addedFunds` of collateral.
+ * @param addedFunds - the amount of collateral being added to the market maker as liquidity
+ * @param poolBalances - the market maker's balances of outcome tokens
+ */
+export const calcAddFundingDepositedAmounts = (
+  addedFunds: bigint,
+  poolBalances: bigint[]
+): bigint[] => {
+  if (poolBalances.some(poolBalance => poolBalance === BigInt(0))) {
+    throw new Error(
+      'Invalid Pool Balances - you must provide a distribution hint for the desired weightings of the pool'
+    );
+  }
+
+  const poolWeight = poolBalances.reduce((a, b) => (a > b ? a : b));
+
+  const depositAmounts = poolBalances.map(h => (addedFunds * h) / poolWeight);
+
+  return depositAmounts;
+};
+/**
+ * Computes the distribution hint that should be used for setting the initial odds to `initialOdds`
+ * @param initialOdds - an array of numbers proportional to the initial estimate of the probability of each outcome
+ */
+export const calcDistributionHint = (initialOdds: number[]): bigint[] => {
+  const allEqual = initialOdds.every(initialOdd => initialOdd === initialOdds[0]);
+  if (allEqual) {
+    return [];
+  }
+  const initialOddsBig = initialOdds.map(x => new Big(x));
+  const product = initialOddsBig.reduce((a, b) => a.mul(b));
+
+  const distributionHint = initialOddsBig
+    .map(o => product.div(o))
+    .map(x => x.mul(1000000).round())
```

Comment:
What's that `1000000` ? The amount of decimals?

Reply:
Looks like it. This code isn't being used but might be useful for market creation. So we need to review this code again when building market creation feature.

Reply:
maybe delete it till then ? OR as least leave a comment there?

Also as this code is being copied from elsewhere, can we add a comment with a link to that on the top of the file?